Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -347,5 +378,125 @@ private bool TryInitializeCorrelationHelperIfNotInitialized()
+            /// </summary>
+            internal void OnBeginRequest_IdRequest(HttpContext context)
+            {
+                this.TagRequest(context);
+            }
+
+            /// <summary>
+            /// OnEndRequest - Should this request be logged?
+            /// Will compare a request id against a hash-set of known requests.
+            /// If this request is not known, add it to hash-set and return true (safe to log).
+            /// If this request is known, return false (do not log twice).
+            /// Additional requests with the same id will return false.
+            /// </summary>
+            internal bool OnEndRequest_ShouldLog(HttpContext context)
+            {
+                var headers = context.Request.Headers;
+
+                var rootRequestId = headers[HeaderRootRequestId];
+                if (rootRequestId != null)
+                {
+                    if (!this.IsRequestKnown(rootRequestId))
+                    {
+                        // doesn't exist add to A and return true;
+                        this.AddRequestToDictionary(rootRequestId);
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            /// <summary>
+            /// Tag new requests.
+            /// Transfer Ids to parent requests.
+            /// </summary>
+            private string TagRequest(HttpContext context)
+            {
+                var newId = Guid.NewGuid().ToString();
+
+                var headers = context.Request.Headers;
+
+                if (headers[HeaderRootRequestId] == null)
+                {
+                    headers[HeaderRootRequestId] = newId;
+                }
+
+#if DEBUG
+                // additional ids help developer watch request hierarchy while debugging.
+                if (headers[HeaderRequestId] != null)
+                {
+                    headers[HeaderParentRequestId] = headers[HeaderRequestId];
+                }
+
+                headers[HeaderRequestId] = newId;
+#endif
+
+                return newId;
+            }
+
+            /// <summary>
+            /// Has this request been tracked.
+            /// </summary>
+            private bool IsRequestKnown(string requestId)
+            {
+                return activeRequestsA.TryGetValue(requestId, out bool valueA) || activeRequestsB.TryGetValue(requestId, out bool valueB);
+            }
+
+            /// <summary>
+            /// Track this requestId.
+            /// </summary>
+            private void AddRequestToDictionary(string requestId)
+            {
+                // TODO: LOCK
+                if (activeRequestsA.Count >= MAXSIZE)
+                {
+                    activeRequestsB = activeRequestsA;
+                    activeRequestsA = new ConcurrentDictionary<string, bool>(System.Environment.ProcessorCount, MAXSIZE);
+                }
+
+                activeRequestsA.TryAdd(requestId, false);
```

Comment:
Do you ever remove items from the dictionary?

Reply:
The intention was not to because the removal was adding extra overhead.

Reply:
Typically there are a few requests running in parallel. Let's say up to 100. It may be cheaper to add/remove this hundred than insert into very big hashtable. When I meant never delete - I meant never delete expired IDs, drop the full hashtable instead.

As you mentioned - it will be great to run some small synthetic tests to understand perf implications of both approaches.

Reply:
When I started this, I wrote a small synthetic test of multiple threads adding and removing entries into a single concurrentdictionary and it had a SIGNIFICANT cpu overhead.

Trusting that we can determine when a request ends requires a few assumptions, and I've been able to find antidotal examples that disprove each assumption. I would be happy to discuss further.